### PR TITLE
macros: improve error massages

### PR DESCRIPTION
**Pull request by taiki-e**
## Motivation

https://github.com/dtolnay/syn/issues/575

## Solution

Using [`syn::Error::new_spanned`](https://docs.rs/syn/0.15.43/syn/struct.Error.html#method.new_spanned).

### CODE REVIEW

File: tokio-macros/src/lib.rs [hunk 1]
```diff
@@ -56,17 +55,15 @@ pub fn main(args: TokenStream, item: TokenStream) -> TokenStream {
...
+            .to_compile_error()
+            .into();
     } else if !input.decl.inputs.is_empty() {
-        let tokens = quote_spanned! { input.span() =>
-          compile_error!("the main function cannot accept arguments");
-        };
-
-        return TokenStream::from(tokens);
+        let msg = "the main function cannot accept arguments";
+        return syn::Error::new_spanned(&input.decl.inputs, msg)
```
Comment (taiki-e) [inline tokio-macros/src/lib.rs]: Before:
```text
error: the main function cannot accept arguments
  --> tokio\examples\echo.rs:34:1
   |
34 | async fn main(foo: Foo) -> Result<(), Box<dyn Error>> {
   | ^^^^^
```
After:
```text
error: the main function cannot accept arguments
  --> tokio\examples\echo.rs:34:15
   |
34 | async fn main(foo: Foo) -> Result<(), Box<dyn Error>> {
   |               ^^^^^^^^
```

File: tokio-macros/src/lib.rs [hunk 2]
```diff
@@ -76,7 +73,12 @@ pub fn main(args: TokenStream, item: TokenStream) -> TokenStream {
             match ident.to_string().to_lowercase().as_str() {
                 "multi_thread" => runtime = RuntimeType::Multi,
                 "single_thread" => runtime = RuntimeType::Single,
-                name => panic!("Unknown attribute {} is specified", name),
+                name => {
+                    let msg = format!("Unknown attribute {} is specified", name);
+                    return syn::Error::new_spanned(ident, msg)
```
Comment (taiki-e) [inline tokio-macros/src/lib.rs]: Before
```text
error: custom attribute panicked
  --> tokio\examples\echo.rs:33:1
   |
33 | #[tokio::main(foo)]
   | ^^^^^^^^^^^^^^^^^^^
   |
   = help: message: Unknown attribute foo is specified
```
After:
```text
error: Unknown attribute foo is specified
  --> tokio\examples\echo.rs:33:15
   |
33 | #[tokio::main(foo)]
   |               ^^^
```

File: tokio-macros/src/lib.rs [hunk 3]
```diff
@@ -56,17 +55,15 @@ pub fn main(args: TokenStream, item: TokenStream) -> TokenStream {
     let attrs = &input.attrs;
 
     if input.asyncness.is_none() {
-        let tokens = quote_spanned! { input.span() =>
-          compile_error!("the async keyword is missing from the function declaration");
-        };
-
-        return TokenStream::from(tokens);
+        let msg = "the async keyword is missing from the function declaration";
+        return syn::Error::new_spanned(input.decl.fn_token, msg)
```
Comment (taiki-e) [inline tokio-macros/src/lib.rs]: There is no change (perhaps the clearest is to highlight `fn`.).
```text
error: the async keyword is missing from the function declaration
  --> tokio\examples\echo.rs:34:1
   |
34 | fn main() -> Result<(), Box<dyn Error>> {
   | ^^
```

File: tokio-macros/src/lib.rs [hunk 4]
```diff
@@ -126,26 +128,23 @@ pub fn test(_attr: TokenStream, item: TokenStream) -> TokenStream {
 
     for attr in attrs {
         if attr.path.is_ident("test") {
-            let tokens = quote_spanned! { input.span() =>
-                compile_error!("second test attribute is supplied");
-            };
-
-            return TokenStream::from(tokens);
+            let msg = "second test attribute is supplied";
+            return syn::Error::new_spanned(&attr, msg)
```
Comment (taiki-e) [inline tokio-macros/src/lib.rs]: There is no change (this should be fine.).
```text
error: second test attribute is supplied
  --> tokio\tests\buffered.rs:14:1
   |
14 | #[test]
   | ^^^^^^^
```

File: tokio-macros/src/lib.rs (hunk 3 above)
Comment (taiki-e) [reply tokio-macros/src/lib.rs]: Note that the original version highlights the entire function when procmacro2_semver_exempt is active.

**Comment by davidbarsky** [comment]
This looks great! This could maybe be added in a followup PR, but how do you feel about using https://github.com/dtolnay/trybuild/ to test the error messages?

**Comment by taiki-e** [comment]
@davidbarsky 
> This could maybe be added in a followup PR, but how do you feel about using https://github.com/dtolnay/trybuild/ to test the error messages?

I think it's nice that the library that provides proc-macro does UI testing.
I've never used trybuild yet (I am using [compiletest-rs](https://github.com/laumann/compiletest-rs)), but it seems [easier to use](https://github.com/dtolnay/tt-call/blob/master/tests/compiletest.rs) than compiletest-rs (compiletest-rs [has many problems](https://github.com/dtolnay/trybuild/issues/13#issuecomment-507077406)...).

**Pull request merged**